src/common/history.c

```diff
@@ -67,6 +67,8 @@ void dt_history_delete_on_image_ext(int32_t imgid, gboolean undo)
     dt_history_snapshot_undo_create(hist->imgid, &hist->before, &hist->before_history_end);
   }
 
+  dt_pthread_mutex_lock(&darktable.db_insert);
```

Comment:
I think this is really good here. Makes much more sense than the lock in develop.c _dev_merge_history